src/cfddlc_transactions.cpp

```diff
@@ -763,23 +471,53 @@ uint32_t DlcManager::GetTotalInputVSize(const std::vector<TxIn>& inputs) {
   return total_size;
 }
 
-bool DlcManager::IsDustOutput(const TxOut& output, uint32_t fee_rate,
-                              uint32_t extra_size) {
-  TxOutReference txout_ref(output);
-  size_t nSize = txout_ref.GetSerializeSize();
-  std::vector<unsigned char> witnessprogram;
+bool DlcManager::IsDustOutput(const TxOut& output) {
+  return output.GetValue() < DUST_LIMIT;
+}
 
-  if (txout_ref.GetLockingScript().IsWitnessProgram()) {
-    // sum the sizes of the parts of a transaction input
-    // with 75% segwit discount applied to the script size.
-    nSize += (32 + 4 + 1 + (107 / WITNESS_SCALE_FACTOR) + 4);
-  } else {
-    nSize += (32 + 4 + 1 + 107 + 4);  // the 148 mentioned above
+static uint32_t GetInputsWeight(const std::vector<TxInputInfo>& inputs_info) {
+  uint32_t total = 0;
+  for (auto input_info : inputs_info) {
+    auto script = input_info.input.GetUnlockingScript();
+    auto script_size = script.IsEmpty() ? 0 : script.GetData().GetDataSize();
+    total += 164 + 4 * script_size + input_info.max_witness_length;
+  }
+
+  return total;
+}
+
+std::tuple<TxOut, uint64_t, uint64_t> DlcManager::GetChangeOutputAndFees(
+    const PartyParams& params, uint64_t fee_rate, Amount option_premium,
+    Address option_dest) {
+  auto inputs_size = GetInputsWeight(params.inputs_info);
+  auto change_size = params.change_script_pubkey.GetData().GetDataSize();
+  double fund_weight =
+      (FUND_TX_BASE_WEIGHT / 2 + inputs_size + change_size * 4 + 36);
+  if (option_premium.GetSatoshiValue() > 0) {
+    if (option_dest.GetAddress() == "") {
+      throw CfdException(
+          CfdError::kCfdIllegalArgumentError,
+          "An destination address for the premium is required when the option "
+          "premium amount is greater than zero.");
+    }
+    fund_weight =
+        36 + option_dest.GetLockingScript().GetData().GetDataSize() * 4;
+  }
+  auto fund_fee = ceil(fund_weight / 4) * fee_rate;
+  double cet_weight = (CET_BASE_WEIGHT / 2 +
+                       params.final_script_pubkey.GetData().GetDataSize() * 4);
+  auto cet_fee = ceil(cet_weight / 4) * fee_rate;
+  auto fund_out = params.collateral + fund_fee + cet_fee;
+  if (params.input_amount < fund_out) {
```

Comment:
I think you should also consider option_premium in this conditional statement.
```suggestion
  if (params.input_amount < (fund_out + option_premium)) {
```

Reply:
Thanks! Fixed